### PR TITLE
Improve the `DataRowCollectionEquivalency` code coverage

### PR DESCRIPTION
**Pull request by sdelarosbil**
Added 4 tests and one static member to DataSpec.

Most of them had to do with edge cases relating to primary keys in the corresponding match mode. I wasn't able to find how you could cover the Equals method on the ConpundKey, but the rest should be covered.

Part of #1823

### CODE REVIEW

File: Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs [hunk 1]
```diff
@@ -32,6 +32,105 @@ public void When_data_tables_are_both_null_equivalence_test_should_succeed()
         ((DataTable)null).Should().BeEquivalentTo(null);
     }
 
+    [Fact]
+    public void When_row_match_mode_is_invalid_it_should_fail()
+    {
+        // Arrange
+        var typedDataSet1 = CreateDummyDataSet<TypedDataSetSubclass>();
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs:39]: 🤔 Do you really need two datasets here? If not, remove one to not obfuscate the test. If you do need it, please use `subject` and `expectation` as the variable names

File: Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs [hunk 2]
```diff
@@ -32,6 +32,105 @@ public void When_data_tables_are_both_null_equivalence_test_should_succeed()
...
+        action.Should().Throw<XunitException>().WithMessage(
+            "Unknown RowMatchMode *when trying to compare *");
+    }
+
+    [Theory]
+    [MemberData(nameof(EmptyPrimaryKeys))]
+    public void When_row_match_mode_is_primary_key_without_primary_key_it_should_fail(DataColumn[] emptyPrimaryKey)
+    {
+        // Arrange
+        var typedDataSet1 = CreateDummyDataSet<TypedDataSetSubclass>(includeRelation: false);
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs:58]: 🤔 Do you really need two datasets here? If not, remove one to not obfuscate the test. If you do need it, please use subject and expectation as the variable names.

File: Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs [hunk 3]
```diff
@@ -32,6 +32,105 @@ public void When_data_tables_are_both_null_equivalence_test_should_succeed()
...
+        var typedDataSet2 = new TypedDataSetSubclass(typedDataSet1);
+
+        var dataTable1 = typedDataSet1.ToUntypedDataSet().Tables["TypedDataTable1"];
+        var dataTable2 = typedDataSet2.ToUntypedDataSet().Tables["TypedDataTable1"];
+
+        dataTable1.PrimaryKey = emptyPrimaryKey;
+
+        // Act
+        Action action = () => dataTable1.Should().BeEquivalentTo(dataTable2, options => options.UsingRowMatchMode(RowMatchMode.PrimaryKey));
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs:67]: 🔧 This line is too long. Please the reformat feature of your IDE so it uses the `.editorconfig` correctly.

File: Tests/FluentAssertions.Equivalency.Specs/DataSpecs.cs [hunk 4]
```diff
@@ -586,6 +586,12 @@ public enum ChangeType
         Removed,
     }
 
+    public static IEnumerable<object[]> EmptyPrimaryKeys => new List<object[]>
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataSpecs.cs:589]: 🔧 Please keep this one close/under the test that uses it.

File: Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs [hunk 5]
```diff
@@ -32,6 +32,105 @@ public void When_data_tables_are_both_null_equivalence_test_should_succeed()
...
+
+        // Assert
+        action.Should().Throw<XunitException>().WithMessage(
+            "*Table *containing *does not have a primary key. RowMatchMode.PrimaryKey cannot be applied.*");
+    }
+
+    [Fact]
+    public void When_primary_key_type_differ_in_primary_key_match_mode_it_should_fail()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs:75]:
```suggestion
    public void When_primary_key_types_do_not_match_it_should_throw()
```

File: Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs [hunk 6]
```diff
@@ -32,6 +32,105 @@ public void When_data_tables_are_both_null_equivalence_test_should_succeed()
...
+        // Assert
+        action.Should().Throw<XunitException>().WithMessage(
+            "*Subject and expectation primary keys of table containing *do not have the same schema and cannot be compared. " +
+            "RowMatchMode.PrimaryKey cannot be applied.*");
+    }
+
+    [Theory]
+    [InlineData(1)]
+    [InlineData(3)]
+    public void When_rows_with_different_primary_key_in_primary_key_match_mode_it_should_fail(int nbrRowsToEdit)
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs:102]: 🔧 This is not grammatically correct. Would you mind to rephrase.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
Thanks for the improvement. I have some (mostly minor) suggestions

**Review by dennisdoomen — APPROVED** [review]

File: Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs (hunk 3 above)
Comment (jnyrup) [reply Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs]: FYI, VS does not support `max_line_length` https://github.com/dotnet/roslyn/issues/15406
But it seems we can use Roslynator to enforce it https://github.com/JosefPihrt/Roslynator/blob/main/docs/analyzers/RCS0056.md to make it IDE/editor agnostic.

File: Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs [hunk 7]
```diff
@@ -32,6 +33,112 @@ public void When_data_tables_are_both_null_equivalence_test_should_succeed()
...
+        // Assert
+        action.Should().Throw<XunitException>().WithMessage(
+            "*Subject and expectation primary keys of table containing *do not have the same schema and cannot be compared. " +
+            "RowMatchMode.PrimaryKey cannot be applied.*");
+    }
+
+    [Theory]
+    [InlineData(1)]
+    [InlineData(3)]
+    public void When_primary_key_of_rows_differ_it_should_fail(int nbrRowsToEdit)
```
Comment (jnyrup) [inline Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs:109]: I would prefer duplicating this test to avoid conditional logic like `for` and `if`.

File: Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs [hunk 8]
```diff
@@ -32,6 +33,112 @@ public void When_data_tables_are_both_null_equivalence_test_should_succeed()
...
+        // Assert
+        action.Should().Throw<XunitException>().WithMessage(
+            "*Table *containing *does not have a primary key. RowMatchMode.PrimaryKey cannot be applied.*");
+    }
+
+    public static IEnumerable<object[]> EmptyPrimaryKeys => new List<object[]>
+    {
+        new object[] { null },
+        new object[] { new DataColumn[] { } }
+    };
```
Comment (jnyrup) [inline Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs:78]:
```suggestion
    public static TheoryData<DataColumn[]> EmptyPrimaryKeys => new()
    {
        null,
        new DataColumn[] { }
    };
```

**Review by jnyrup — CHANGES_REQUESTED** [review]

**Review by jnyrup — APPROVED** [review]
